cmd/registry/scoring/score.go

```diff
@@ -0,0 +1,366 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scoring
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/apigee/registry/cmd/registry/core"
+	"github.com/apigee/registry/cmd/registry/patch"
+	"github.com/apigee/registry/cmd/registry/patterns"
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/log"
+	"github.com/apigee/registry/rpc"
+	"github.com/apigee/registry/server/registry/names"
+	"google.golang.org/protobuf/proto"
+)
+
+func FetchScoreDefinitions(
+	ctx context.Context,
+	client connection.Client,
+	resource patterns.ResourceName) ([]*rpc.ScoreDefinition, error) {
+	definitions := make([]*rpc.ScoreDefinition, 0)
+
+	project := fmt.Sprintf("%s/locations/global", resource.Project())
+	artifact, err := names.ParseArtifact(fmt.Sprintf("%s/artifacts/-", project))
+	if err != nil {
+		return nil, err
+	}
+	listFilter := fmt.Sprintf("mime_type == %q", patch.ScoreDefinitionMimeType)
+	err = core.ListArtifacts(ctx, client, artifact, listFilter, true,
+		func(artifact *rpc.Artifact) error {
+
+			definition := &rpc.ScoreDefinition{}
+			if err := proto.Unmarshal(artifact.GetContents(), definition); err != nil {
+				return err
+			}
+
+			// Check if ScoreDefinition.TargetResource matches with the supplied resource
+			err := matchResourceWithTarget(definition.GetTargetResource(), resource, project)
+			if err != nil {
+				return err
+			}
+
+			definitions = append(definitions, definition)
+			return nil
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return definitions, nil
+}
+
+func matchResourceWithTarget(targetPattern *rpc.ResourcePattern, resourceName patterns.ResourceName, project string) error {
+	targetPatternName, err := patterns.ParseResourcePattern(fmt.Sprintf("%s/%s", project, targetPattern.GetPattern()))
+	if err != nil {
+		return err
+	}
+
+	// Check if targetPattern and resourceName match in type
+	if fmt.Sprintf("%T", targetPatternName) != fmt.Sprintf("%T", resourceName) {
+		return fmt.Errorf("resource %s doesn't match target pattern %v", resourceName.String(), targetPattern)
+	}
+
+	// Check if the individual entities match
+	switch targetPatternName.(type) {
+	case patterns.SpecName:
+		// type casting because we need access to "Name" field of patterns.SpecName to compare specific IDs
+		specPattern := targetPatternName.(patterns.SpecName)
+		specResource := resourceName.(patterns.SpecName)
```

Comment:
How certain are we that these are always the same type? The resource name isn't type checked, and will panic if it doesn't match.

Reply:
Done. I will do the error message refactor in a separate PR.  I think that can be done more broadly across the validation functions and it's tests.